src/Limiters/limiter.jl

```diff
@@ -0,0 +1,119 @@
+"""
+    quasimonotone_limiter!(ρq, ρ, min_ρq, max_ρq; rtol)
+
+Arguments:
+- `ρq`: tracer density Field, where `q` denotes tracer concentration (or mixing
+  ratio) per unit mass
+- `ρ: fluid density Field
+- `min_ρq: Array of min(ρq) per element
+- `max_ρq`: Array of max(ρq) per element
+- `rtol`: relative tolerance needed to solve element-wise optimization problem
+
+This limiter is inspired by the one presented in Guba et al [GubaOpt2014](@cite).
+In the reference paper, it is denoted by OP1, and is outlined in eq.s (37)-(40).
+Quasimonotone here is meant to be monotone with respect to the spectral element
+nodal values. This limiter involves solving a constrained optimization problem
+(a weighted least square problem up to a fixed tolerance denoted by `rtol`)
+that is completely local to each element.
+As in HOMME, the implementation idea here is the following: we need to find a grid
+field which is closest to the initial field (in terms of weighted sum), but satisfies
+the min/max constraints. So, first we find values which do not satisfy constraints
+and bring these values to a closest constraint. This way we introduce some mass
+change (`mass_change`), so, we redistribute `mass_change` in the way that the
+l2 error is smallest.This redistribution might violate constraints thus, we do
+a few iterations.
+"""
+function quasimonotone_limiter!(
+    ρq::Fields.Field,
+    ρ::Fields.Field,
+    min_ρq,
+    max_ρq;
+    rtol,
+)
+    space = axes(ρ)
+    FT = Spaces.undertype(space)
+    Nq = Spaces.Quadratures.degrees_of_freedom(Spaces.quadrature_style(space))
+    n_elems = length(min_ρq)
+    q = Array{FT, 2}(undef, Nq, Nq)
+    ρ_wJ = Array{FT, 2}(undef, Nq, Nq)
+
+    # Compute ρ's and ρq's masses over an element e
+    for e in 1:n_elems
+        ρ_e_mass = sum(Fields.slab(ρ, e))
+        ρq_e_mass = sum(Fields.slab(ρq, e))
+
+        # This should never happen, but if it does, don't limit
+        if ρ_e_mass <= 0
+            continue
+        end
+
+        # Relax constraints to ensure limiter has a solution:
+        # This is only needed if runnign with the SSP CFL>1 or
+        # due to roundoff errors.
+        if ρq_e_mass < min_ρq[e] * ρ_e_mass
+            min_ρq[e] = ρq_e_mass / ρ_e_mass
+        end
+
+        if ρq_e_mass > max_ρq[e] * ρ_e_mass
+            max_ρq[e] = ρq_e_mass / ρ_e_mass
+        end
+
+        local_geometry_slab = Fields.slab(space.local_geometry, e)
+
+        q = parent(Fields.slab(ρq, e)) ./ parent(Fields.slab(ρ, e))
```

Comment:
for performance it would be better to pull out the required views once,  `\rho_q_slab = ...`,  `\rho_slab  = ...` as you are constructing them multiple times.